tt_metal/api/program.hpp

```diff
@@ -0,0 +1,162 @@
+// SPDX-FileCopyrightText: © 2024 Tenstorrent Inc.
+//
+// SPDX-License-Identifier: Apache-2.0
+
+#pragma once
+
+#include <string_view>
+#include "types.hpp"
+
+#include "tt_metal/impl/kernels/kernel_types.hpp"
+#include "tt_metal/impl/buffers/circular_buffer_types.hpp"
+
+//==================================================
+//                  PROGRAM MANAGEMENT
+//==================================================
+
+namespace tt::tt_metal{
+namespace v1 {
+
+/**
+ * @brief Creates a Program object, which bundles kernels, circular buffers, and semaphores for execution on the device.
+ *
+ * @return Program handle to the created program.
+ */
+Program CreateProgram();
+
+
+/**
+ * @brief Creates a data movement or compute kernel and adds it to the program.
+ *
+ * @param program The program to which this kernel will be added.
+ * @param file_name Path to the kernel source file.
+ * @param core_spec Specifies the cores on which the kernel will be placed.
+ * @param config DataMovementConfig for the kernel.
+ * @return KernelHandle representing the kernel ID.
+ */
+KernelHandle CreateKernel(
```

Comment:
What is the error handling policy for these APIs?

**Example situation:**
Assuming a Program should not be accessed by anyone after it is moved to device.
What if I send a program for execution to a device and later try to call CreateKernel on the same program handle?

@pgkeller @eyonland

Reply:
@pgkeller, would it make more sense to just return an invalid KernelHandle instead of throwing an exception?

Reply:
We need a deep dive on error handling, I'm not thrilled w/ what we have.

Today we do a lot of "validate as you go" and we're going to move to more lazy validation which means errors can happen apart from the bogus API call (for example, we'll learn a kernel is invalid when we compile the program).

Our testing suffers today because we print "fatal error" for tests that trigger errors intentionally and catch the exceptions (yuck).  Maybe this could be handled by a back-door test API that disables error reporting.

A kernel compilation error currently throws an exception which pollutes the error message and makes kernel debugging harder (if I have a typo in a datamovement kernel I really do not care about the call stack to the build fn).

Need to decide on a consistent approach here.